feat(typography): create typography component FE-3068, FE-3002

jamime opened the pull request:
### Proposed behaviour

- Create `Typography`,`List` and `ListItem` components.
- Refactor palette so all colors can be used by the component directly.
- Add `color` util which can be used by all components.
- Update `box` to use new `color` util.
- Add MDX documentation for `Typography` and `Color`.
- Add eslint rule to prevent users using styled-system color directly.
- Add missing `backgroundColor` to Color definition.


> **BREAKING CHANGE:** The `atOpacity` function has been removed from the theme palette. To use this utility you should import it directly `import atOpacity from “carbon-react/lib/style/utils/at-opacity`.

> **BREKAING CHANGE:** The file “carbon-react/lib/style/index” has been removed.
If you want to use `addOpacity` you can use `atOpacity` directly.
If you want to use `blackAtOpacity` you can use `theme.palette.blackOpacity` or you can use `atOpacity` directly.
Please note that `blackAtOpacity` previously returned a white opacity you can use `theme.palette.whiteOpacity` to keep this behaviour.
If you want to use `generatePalette` you can import it directly `import generatePalette from “carbon-react/lib/style/palette`.

![image](https://user-images.githubusercontent.com/2328042/96864926-1ca6aa80-1461-11eb-981f-974e8413ea90.png)


### Current behaviour

- No typography components
- Duplicated code around `palette` and `atOpacity`
- Missing `backgroundColor` in Color definition.

### Checklist
<!-- Each PR should include the following -->

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Additional context
FE-3068, FE-3002


### Testing instructions

See the new MDX pages for Color and Typography. No need for cypress tests, there are extensive unit tests.

sjohnson-sage commented on docs/colors.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+
+# Colors
+
+Some componets such as <LinkTo kind='Design System/Typography'>Typography</LinkTo> have a `color`, `bg`/`backgroundColor` and `opacity` props, this functionality is provided by <a href="https://styled-system.com/api/#color">`styled-system/color`</a>.
```
```diff
components
componets
```

sjohnson-sage commented on src/components/typography/list.component.js:
```diff
@@ -0,0 +1,39 @@
+
+const ListItem = ({ children, ...props }) => (
+  <Typography
+    as='li' variant='p'
```
Really minor but I'd put these props on separate lines as above. I'll hurry up and get prettier in next pioneer Friday so we don't even have to think about this!

sjohnson-sage commented on src/components/typography/typography.component.js:
```diff
@@ -0,0 +1,211 @@
+    textTransform: textTransform || getTransform(variant),
+    textDecoration: textDecoration || getDecoration(variant),
+    lineHeight: lineHeight || getLineHeight(variant),
+    defaultMargin: variant === 'p' ? '0 0 16px 0' : '0'
```
Minor again but you don't actually need the last 0 for the margin here
```suggestion
    defaultMargin: variant === 'p' ? '0 0 16px' : '0'
```

That will set left and right margins to 0

sjohnson-sage reviewed (commented):
Looks really good, just a few pretty minor comments from me then happy to approve 👍

DlgSHi commented on docs/colors.stories.mdx:
```diff
@@ -0,0 +1,50 @@
+import LinkTo from '@storybook/addon-links/react';
+
```
unnecessary empty line :)

mkrds commented on docs/colors.stories.mdx:
```diff
@@ -0,0 +1,47 @@
+
+## Palette Colors
+
+Using the name of a palette color ensures if we update the theme your components will also be updated, for this reason we advise not using a CSS sting.
```
```suggestion
Using the name of a palette color ensures if we update the theme your components will also be updated, for this reason we advise not using a CSS string.
```

jamime commented:
@harpalsingh found some missing props from the variant enum in the manual table. I'll add them.

jamime commented:
I've added the missing props from the variant enum in the manual table. I also added `backgroundColor` to the typescript definitions and to the styled-system prop table.

carbonci commented:
:tada: This PR is included in version 47.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-react/v/47.0.0)
- [GitHub release](https://github.com/Sage/carbon/releases/tag/v47.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
